feat: add profile gst_capi face_detection

vli11 opened the pull request:
## PR Checklist
<!-- Please check if your PR fulfills the following requirements: -->

- [x] Added label to the Pull Request for easier discoverability and search
- [x] Commit Message meets guidelines as indicated in the URL https://github.com/intel-retail/automated-self-checkout/blob/main/CONTRIBUTING.md
- [x] Every commit is a single defect fix and does not mix feature addition or changes
- [ ] Unit Tests have been added for new changes
- [ ] Updated Documentation as relevant to the changes
- [ ] All commented code has been removed
- [ ] If you've added a dependency, you've ensured license is compatible with repository license and clearly outlined the added dependency.
- [ ] PR change contains code related to security
- [ ] PR introduces changes that breaks compatibility with other modules (If YES, please provide details below)

## What are you changing?
add feature: gst capi face_detection profile

## Issue this PR will close

close: #**issue_number**

## Anything the reviewer should know when reviewing this PR?

## Test Instructions if applicable
1. make build-gst-capi
2. ./camera-simulator/camera-simulator.sh
3. PIPELINE_PROFILE="face_detection" RENDER_MODE=1 sudo -E ./run.sh --workload ovms --inputsrc rtsp://127.0.0.1:8554/camera_0 --platform core

after above 3 commands, you should see video streaming result of inferencing popped up.

## If the there are associated PRs in other repositories, please link them here (i.e. intel-retail/automated-self-checkout )

vli11 committed:
feat: add profile gst_capi face_detection

Signed-off-by: Valina Li <valina.li@intel.com>

Brian-McGinn commented on configs/opencv-ovms/gst_capi/Makefile:
```diff
@@ -0,0 +1,28 @@
+	echo "Building GStreamer OVMS C-API Client HTTPS_PROXY=${HTTPS_PROXY} HTTP_PROXY=${HTTP_PROXY}"
+	# Build C-API for optimized distributed architecture. Includes GST for HWA media	
+	rm -vrf ovms.tar.gz 
+	wget -O ovms.tar.gz https://github.com/openvinotoolkit/model_server/releases/download/v2023.0/ovms_ubuntu22.tar.gz
```
I believe this is unused now that we are pulling the latest OVMS image from Dockerhub

vli11 replied:
we need the ovms/lib in the container to build the face_detection c++ code @brian-intel

Brian-McGinn commented on configs/opencv-ovms/gst_capi/get-media-codec.sh:
```diff
@@ -0,0 +1,21 @@
+   	show_help
+	exit 0
+else
+    is_avc=`gst-discoverer-1.0 $1 | grep -i h.264 | wc -l`
```
Is it possible to move this logic into the launch-pipeline.sh. It seems a bit overkill to have a separate script for a single function.

vli11 committed:
fix shellcheck errors

Signed-off-by: Valina Li <valina.li@intel.com>

vli11 committed:
add face_detection.env file

Signed-off-by: Valina Li <valina.li@intel.com>

vli11 committed:
consolidate script files

Signed-off-by: Valina Li <valina.li@intel.com>

vli11 committed:
add face_detection run script

Signed-off-by: Valina Li <valina.li@intel.com>

ejlee3 commented on configs/opencv-ovms/gst_capi/Dockerfile.ovms-capi-gst:
```diff
@@ -0,0 +1,95 @@
+ENV LD_LIBRARY_PATH=/usr/lib/x86_64-linux-gnu/:/usr/local/lib/x86_64-linux-gnu/:/usr/local/lib:/ovms/lib:$LD_LIBRARY_PATH
+COPY gst_capi/pipelines /home/intel/gst-ovms/pipelines
+COPY models/2022 /home/intel/gst-ovms/models
+COPY gst_capi/get-media-codec.sh /home/intel/gst-ovms/get-media-codec.sh
```
this line is breaking the build because the get-media-codec.sh doesn't exist?

vli11 committed:
clean dockerfile

Signed-off-by: Valina Li <valina.li@intel.com>

ejlee3 reviewed (approved):
LGTM - tested on my linux nuc. One thing: is there a good way to bring things down and clean up the services?

vli11 committed:
move building of binary into runtime

Signed-off-by: Valina Li <valina.li@intel.com>

jim-wang-yutsung commented on Makefile:
please also add the `clean-gst-capi` make target and add that target as dependency of clean-profile-launcher as well

jim-wang-yutsung commented on configs/opencv-ovms/gst_capi/pipelines/face_detection/main.cpp:
```diff
@@ -0,0 +1,1019 @@
+        std::vector<int> input_shape = getModelInputShape();
+        int network_h =  input_shape[2];
+        int network_w =  input_shape[3];
+        // printf("Network %f %f numDets %d outputDims: %d imageId: %f label: %f  \n",
+        //     network_h, network_w, numberOfDetections, objectSize, outData[0 * objectSize + 0], outData[0 * objectSize + 1]);
```
remvoe

jim-wang-yutsung commented on configs/opencv-ovms/gst_capi/pipelines/face_detection/main.cpp:
```diff
@@ -0,0 +1,1019 @@
+
+            float confidence = outData[i * objectSize + 2];
+
+            //printf("Confidence found: %f\n", confidence);
```
remove

jim-wang-yutsung commented on configs/opencv-ovms/gst_capi/pipelines/face_detection/main.cpp:
```diff
@@ -0,0 +1,1019 @@
+            //printf("Confidence found: %f\n", confidence);
+
+            if (confidence > confidence_threshold ) {
+                //printf("Confidence found: %f\n", confidence);
```
remove

jim-wang-yutsung commented on configs/opencv-ovms/gst_capi/pipelines/face_detection/main.cpp:
```diff
@@ -0,0 +1,1019 @@
+        const float x1 = obj.x + obj.width;
+        const float y1 = obj.y + obj.height;
+
+        //printf("--------->coords: %f %f %f %f\n", x0, y0, x1, y1);
```
remove

jim-wang-yutsung commented on configs/opencv-ovms/gst_capi/pipelines/face_detection/main.cpp:
```diff
@@ -0,0 +1,1019 @@
+
+        const float scaler_w = 416.0f/_video_input_width;
+        const float scaler_h = 416.0f/_video_input_height;
+        //std::cout << " Scalers " << scaler_w << " " << scaler_h << std::endl;
+        //std::cout << "xDrawing at " << (int)obj.x*scaler_w << "," << (int)obj.y*scaler_h << " " << (int)(obj.x+obj.width)*scaler_w << " " << (int) (obj.y+obj.height)* scaler_h << std::endl;
```
remvoe

jim-wang-yutsung commented on configs/opencv-ovms/gst_capi/pipelines/face_detection/main.cpp:
```diff
@@ -0,0 +1,1019 @@
+            if (dynamic_cast<const FaceDetection0005*>(objDet) != nullptr)
+	        {
+                resize(img, analytics_frame, cv::Size(inputShape[2], inputShape[3]), 0, 0, cv::INTER_LINEAR);
+                //cv::imwrite("faceresized.jpg", analytics_frame);
```
remove

jim-wang-yutsung commented on configs/opencv-ovms/gst_capi/pipelines/face_detection/main.cpp:
```diff
@@ -0,0 +1,1019 @@
+        outputId = outputCount - 1;
+
+        OVMS_InferenceResponseGetOutput(response, outputId, &outputName1, &datatype1, &shape1, &dimCount1, &voutputData1, &bytesize1, &bufferType1, &deviceId1);
+        // std::cout << "------------>" << tid << " : " << "DeviceID " << deviceId1
+        //  << ", OutputName " << outputName1
+        //  << ", DimCount " << dimCount1
+        //  << ", shape " << shape1[0] << " " << shape1[1] << " " << shape1[2] << " " << shape1[3]
+        //  << ", byteSize " << bytesize1
+        //  << ", OutputCount " << outputCount << std::endl;
```
remove

jim-wang-yutsung commented on configs/opencv-ovms/gst_capi/pipelines/face_detection/main.cpp:
```diff
@@ -0,0 +1,1019 @@
+                total_latency_frames = 0;
+            }
+
+            //saveInferenceResultsAsVideo(img, detectedResultsFiltered);
```
remove

jim-wang-yutsung commented on configs/opencv-ovms/gst_capi/pipelines/face_detection/main.cpp:
```diff
@@ -0,0 +1,1019 @@
+
+int get_running_servers() {
+    char buffer[128];
+    //string cmd = "cids=$(docker ps  --filter=\"name=gst-ovms\" -q -a); cid_count=`echo \"$cids\" | wc -w`; echo $cid_count";
```
remove

jim-wang-yutsung commented on download_models/downloadGSTCAPIModels.sh:
why do we need this script?  isn't the current model-downloader from omz-downloader not working?

vli11 replied:
this is just to keep gst capi model separately, it is cleaner; and in case we want to separate the model download process, its already so.

jim-wang-yutsung commented on download_models/getModels.sh:
```diff
@@ -51,19 +51,21 @@ get_options() {
     echo "Starting open-ovms model download..."
-    $MODEL_EXEC_PATH/downloadOVMSModels.sh $REFRESH_INPUT
+    "$MODEL_EXEC_PATH"/downloadOVMSModels.sh $REFRESH_INPUT
+    echo "Starting gst-capi model download..."
+    "$MODEL_EXEC_PATH"/downloadGSTCAPIModels.sh $REFRESH_INPUT
```
remove if this special script is not needed

jim-wang-yutsung commented on configs/opencv-ovms/scripts/run_face_detection.sh:
```diff
@@ -0,0 +1,80 @@
+TAG=openvino/model_server-capi-gst-ovms:latest
+
+# Need to rebuild binary again since config.json updated, then launch-pipeline.sh
+bash_cmd="cd pipelines && chmod +x build-demos.sh && ./build-demos.sh && cd .. && ./launch-pipeline.sh $PIPELINE_EXEC_PATH $INPUTSRC $USE_ONEVPL $RENDER_MODE $RENDER_PORTRAIT_MODE"
```
if we could, maybe worth investigating change this $INPUTSRC to $inputsrc so that we don't change run_ovms.sh at all

vli11 replied:
I can log an issue since $INPUTSRC to $inputsrc can impact other profile run.

vli11 replied:
https://github.com/intel-retail/automated-self-checkout/issues/291 created

vli11 committed:
fix comments

Signed-off-by: Valina Li <valina.li@intel.com>

Commit:
feat: Use docker to build profile-launcher

Signed-off-by: Antonio Martinez <jose.antonio.martinez.torres@intel.com>

Commit:
addressing PR comments

Signed-off-by: Antonio Martinez <jose.antonio.martinez.torres@intel.com>

Commit:
addressing PR comments

Signed-off-by: Antonio Martinez <jose.antonio.martinez.torres@intel.com>

Commit:
Merge pull request #294 from antoniomtz/build-profile-launcher

feat: Use docker to build profile-launcher

ejlee3 committed:
fix: Add code to download yolov5 FP32-INT8

Closes #296

Signed-off-by: Elizabeth J Lee <elizabeth.j.lee@intel.com>

Commit:
feat: implementation of single container / client-server through profile-launcher (#290)

* feat: expand profile-launcher config yaml to include the single container case
* feat: implementation of single container / client-server through profile-launcher
* fix: update other configuration yaml profile to use the latest profile-launcher config
* feat: add smoke tests script and  makefile target to smoke testing ovms profiles
* feat: add the retry logic in the run_smoke_test.sh shell script
* fix: add grpc-port availability check in smoke test in case the port is busy

Signed-off-by: Jim Wang <yutsung.jim.wang@intel.com>

---------

Signed-off-by: Jim Wang <yutsung.jim.wang@intel.com>

ejlee3 committed:
fix: Address Jim's PR comment

Signed-off-by: Elizabeth J Lee <elizabeth.j.lee@intel.com>

Commit:
fix: updated env vars for pipelines

Signed-off-by: Antonio Martinez <jose.antonio.martinez.torres@intel.com>

Commit:
refact: changed docker image to convert sample videos

Signed-off-by: Antonio Martinez <jose.antonio.martinez.torres@intel.com>

ejlee3 committed:
Merge pull request #297 from ejlee3/fix-yolov5-download

fix: Add code to download yolov5 FP32-INT8

Commit:
Merge pull request #304 from antoniomtz/update-docker-image

refact: changed docker image to convert sample videos

Commit:
docs: Updating README for a QuickStart

Signed-off-by: Antonio Martinez <jose.antonio.martinez.torres@intel.com>

Commit:
removing extra space

Signed-off-by: Antonio Martinez <jose.antonio.martinez.torres@intel.com>

Commit:
Merge pull request #299 from antoniomtz/fix-env-variables

fix: Fixed env variables for pipelines in a different Linux distribution

Commit:
addressing PR comments

Signed-off-by: Antonio Martinez <jose.antonio.martinez.torres@intel.com>

Commit:
Merge pull request #305 from antoniomtz/getting-started-guide

docs: Updating README for a QuickStart

ejlee3 committed:
docs: Update note format to use pymdownx.details, add footer

Signed-off-by: Elizabeth J Lee <elizabeth.j.lee@intel.com>

ejlee3 committed:
fix: Pipelines -> Pipeline

Signed-off-by: Elizabeth J Lee <elizabeth.j.lee@intel.com>

ejlee3 committed:
Merge pull request #307 from ejlee3/update_docs_notes

docs: Update note format to use pymdownx.details, add footer

vli11 committed:
feat: Refactor inputsrc and remove INPUTSRC_TYPE from ovms (#301)

feat: Refactor inputsrc and remove INPUTSRC_TYPE from ovms

Moving gst profile needed env variable manipulations to backend run_gst.sh

Closes: #291
Signed-off-by: Valina Li <valina.li@intel.com>

vli11 committed:
fix comments

Signed-off-by: Valina Li <valina.li@intel.com>

vli11 committed:
moving face_detection binary build to dockerbuild

Signed-off-by: Valina Li <valina.li@intel.com>

vli11 committed:
no need model copy

Signed-off-by: Valina Li <valina.li@intel.com>

vli11 committed:
rebase to main and clean up env

Signed-off-by: Valina Li <valina.li@intel.com>

vli11 committed:
clean up

Signed-off-by: Valina Li <valina.li@intel.com>

vli11 committed:
direct install gstreamer

Signed-off-by: Valina Li <valina.li@intel.com>

vli11 commented:
PR moved to main, closing this one
